src/openforms/payments/tests/test_models.py

```diff
@@ -12,86 +13,110 @@
-
-        payment = SubmissionPayment.objects.create_for(
-            submission, "plugin1", options, amount
-        )
-        self.assertEqual(payment.order_id, 4)
-        self.assertEqual(payment.public_order_id, "000004")
-
-        # check overflow over default 5 digit non-year part
-        SubmissionPaymentFactory.create(order_id=10000000)
+        self.assertEqual(payment.public_order_id, "OF-123456/3")
 
         payment = SubmissionPayment.objects.create_for(
             submission, "plugin1", options, amount
         )
-        self.assertEqual(payment.order_id, 10000001)
-        self.assertEqual(payment.public_order_id, "10000001")
+        self.assertEqual(payment.public_order_id, "OF-123456/4")
 
     @freeze_time("2020-01-01")
-    def test_create_for_with_prefix(self):
+    @patch(
+        "openforms.payments.models.GlobalConfiguration.get_solo",
+        return_value=GlobalConfiguration(payment_order_id_prefix="xyz{year}"),
+    )
+    def test_create_for_with_prefix(self, m: MagicMock):
         amount = Decimal("11.25")
         options = {
             "foo": 123,
         }
-        config = GlobalConfiguration.get_solo()
-        config.payment_order_id_prefix = "xyz{year}"
-        config.save()
 
         submission = SubmissionFactory.create(
             form__payment_backend="plugin1",
             form__payment_backend_options=options,
+            public_registration_reference="OF-123456",
         )
-        # create some existent records
-        SubmissionPaymentFactory.create(order_id=1)
-        SubmissionPaymentFactory.create(order_id=2)
+        # create some pre-existing records
+        SubmissionPaymentFactory.create_batch(size=2)
 
         # create payment with auto-generated order_id
         payment = SubmissionPayment.objects.create_for(
             submission, "plugin1", options, amount
         )
-        self.assertEqual(payment.order_id, 3)
-        self.assertEqual(payment.public_order_id, "xyz2020000003")
+        self.assertEqual(payment.public_order_id, "xyz2020/OF-123456/3")
 
         payment = SubmissionPayment.objects.create_for(
             submission, "plugin1", options, amount
         )
-        self.assertEqual(payment.order_id, 4)
-        self.assertEqual(payment.public_order_id, "xyz2020000004")
+        self.assertEqual(payment.public_order_id, "xyz2020/OF-123456/4")
+
+    @patch(
+        "openforms.payments.models.GlobalConfiguration.get_solo",
+        return_value=GlobalConfiguration(
+            payment_order_id_prefix="", wait_for_payment_to_register=True
+        ),
+    )
+    @patch(
+        "openforms.submissions.public_references.get_random_reference",
+        return_value="OF-123456",
+    )
+    def test_create_for_no_registration(self, m1: MagicMock, m2: MagicMock):
+        amount = Decimal("11.25")
+        options = {
+            "foo": 123,
+        }
+
+        submission = SubmissionFactory.create(completed_not_preregistered=True)
```

Comment:
@SilviaAmAm is this right? a deferred registration because it's not paid yet, will it also not pre-register?

Reply:
The special code in payments was added because I had a test failure where this view:
https://github.com/open-formulieren/open-forms/blob/9223c82e4ac480ac60c1f54c905ef9c76c5a6e1a/src/openforms/payments/views.py#L99-L126

would call `create_for` for a submission without a reference. This was coming from a test related to this view I think, but maybe that test wasn't realistic and should be updated?

Reply:
yes, take a look at that test (and check with @SilviaAmAm ) if that is a representative setup - I just fixed another test in Vasileios' PR that had a wrong setup and caused test failure.